service/src/routes/dips.rs

```diff
@@ -0,0 +1,300 @@
+        let rlp_bytes = STANDARD.decode(self.signed_payload)?;
+        let signed_voucher = SignedIndexingAgreementVoucher::decode(&mut rlp_bytes.as_ref())?;
+
+        Ok(signed_voucher)
+    }
+}
+
+#[derive(SimpleObject, Debug, Clone)]
+pub struct Price {
+    price_per_block: String,
+    chain_id: String,
+    protocol_network: String,
+}
+
+#[derive(Debug)]
+pub struct AgreementQuery {}
+
+#[Object]
+impl AgreementQuery {
+    pub async fn get_agreement_by_id<'a>(
+        &self,
+        ctx: &'a Context<'_>,
+        id: String,
+    ) -> FieldResult<Option<Agreement>> {
+        let store: &Arc<dyn AgreementStore> = ctx.data()?;
+        let id = Uuid::from_str(&id)?;
+
+        match store
+            .get_by_id(id)
+            .await
+            .map_err(async_graphql::Error::from)?
+        {
+            Some(a) => (id, a)
+                .try_into()
+                .map(Some)
+                .map_err(async_graphql::Error::from),
+            None => Ok(None),
+        }
+    }
+
+    pub async fn get_price<'a>(
+        &self,
+        ctx: &'a Context<'_>,
+        protocol_network: String,
+        chain_id: String,
+    ) -> FieldResult<Option<Price>> {
+        let prices: &Vec<Price> = ctx.data()?;
+
+        let p = prices
+            .iter()
+            .find(|p| p.protocol_network.eq(&protocol_network) && p.chain_id.eq(&chain_id));
+
+        Ok(p.cloned())
+    }
+
+    pub async fn get_all_prices<'a>(&self, ctx: &'a Context<'_>) -> FieldResult<Vec<Price>> {
+        let prices: &Vec<Price> = ctx.data()?;
+
+        Ok(prices.clone())
+    }
+}
+
+#[derive(Debug)]
+pub struct AgreementMutation {
+    pub expected_payee: Address,
+    pub allowed_payers: Vec<Address>,
+    pub domain: Eip712Domain,
+    pub cancel_voucher_time_tolerance: Duration,
+}
+
+#[Object]
+impl AgreementMutation {
+    // create_agreements returns the UUID under which the agreement was stored.
+    pub async fn create_agreement<'a>(
+        &self,
+        ctx: &'a Context<'_>,
+        // uuid v7 that this agreement should use.
+        id: String,
+        // data should be the signed voucher, eip712 signed, rlp and base64 encoded.
+        signed_voucher: String,
+    ) -> FieldResult<String> {
+        let store: &Arc<dyn AgreementStore> = ctx.data()?;
+        let uid = Uuid::from_str(&id)?;
+
+        validate_and_create_agreement(
+            store.clone(),
+            &self.domain,
+            uid,
+            &self.expected_payee,
+            &self.allowed_payers,
+            signed_voucher,
+        )
+        .await
+        .map_err(async_graphql::Error::from)?;
+
+        Ok(id)
+    }
+
+    pub async fn cancel_agreement<'a>(
+        &self,
+        ctx: &'a Context<'_>,
+        id: String,
+        // data should be the signed voucher, eip712 signed, rlp and base64 encoded.
+        signed_request: String,
+    ) -> FieldResult<String> {
+        let store: &Arc<dyn AgreementStore> = ctx.data()?;
```

Comment:
Out of curiosity. Why did you choose dynamic dispatch over a generic argument?

Reply:
Normally I find it easier to work with dynamic dispatch, generic arguments tend to propagate through the code very easily, I don't think the perf penalty on something like this is worth the trade-off.